Documentation update on tickless mode porting guide (#1337)
ARMmbed/Handbook: docs/porting/target/tickless/tickless.md

Tickless mode is an optimization mechanism available in RTOS for suspending the SysTick. You can use it in situations when RTOS is idle for multiple ticks, so you can achieve power savings by entering uninterrupted sleep. 

# Reminder on scheduling & sleep modes in Mbed OS

Mbed OS uses the SysTick timer at period of 1ms to process threads' scheduling.
For instance, a system running two threads would see its timing diagram look like :

![](./resources/Normal_Tick.png)

Note how the device never enters deepsleep and wastes cycles in Systick while all thread are asleep.

# Tickless mode

To support tickless mode in Mbed OS, your target needs to meet two requirements:

- Support for Sleep HAL API
- Support for Low Power Ticker HAL API

To enable tickless mode support in Mbed OS, you need to add the `MBED_TICKLESS` macro in the macros option of the target's section in the `targets.json` file.

When the tickless mode is enabled, the device suspends the kernel effectively disabling the systick and sets up the [low power ticker] through [mbed::internal::OsTimer].

> If the target does not feature a low power ticker or has an excessively long wake-up time, you must also set the `tickless-from-us-ticker` to true in the target's section in the `targets.json` file. In such the device will only enter sleep rather than deepsleep.

Reviewer: Didn't we just merge a PR that meant you didn't have to set this if you don't have a low power ticker at all? It's bright enough to realise that it has to use us-ticker now. So I think you only have to set this if you have it and it doesn't work well.

Reviewer: If that PR has been merged, the "if the target does not feature a low power ticker" condition is no longer true.

How about "By default, tickless mode uses the low-power ticker if available. If a target's low-power ticker has an excessively long wake-up time, or other performance issues, you must set `tickless-from-us-ticker` to true in the target's section in the `targets.json` file to make it use the microsecond ticker instead."

Then "If tickless mode uses the microsecond ticker, the device will only enter sleep rather than deepsleep, but will still avoid unnecessary tick calls."

Reviewer: Just checked, and the PR is still pending, but ready for merge:

https://github.com/ARMmbed/mbed-os/pull/12988

# Tickless mode

Tickless mode is an optimization mechanism available in RTOS for suspending the SysTick. You can use it in situations when RTOS is idle for multiple ticks, so you can achieve power savings by entering uninterrupted sleep. 

> ## Reminder on scheduling & sleep modes in Mbed OS

Reviewer 2: ```suggestion
## Scheduling and sleep modes in Mbed OS
```

> Mbed OS uses the SysTick timer at period of 1ms to process threads' scheduling.

Reviewer 2: ```suggestion
Mbed OS uses the SysTick timer in periods of 1ms to process threads' scheduling.
```

> For instance, a system running two threads would see its timing diagram look like :

Reviewer 2: ```suggestion
For instance, a system running two threads would see its timing diagram look like:
```

> ![](./resources/Normal_Tick.png)

Reviewer 2: ```suggestion
![](resources/Normal_Tick.png)
```

> Note how the device never enters deepsleep and wastes cycles in Systick while all thread are asleep.

Reviewer 2: ```suggestion
Note that the device never enters deep sleep and wastes cycles in SysTick while all threads are asleep.
```

Reviewer 2: Are we trying to say "Note that the device doesn't enters deep sleep and doesn't waste cycles in SysTick if all threads are asleep"?

## Tickless mode

To support tickless mode in Mbed OS, your target needs to meet two requirements:

- Support for Sleep HAL API
- Support for either Low Power or Microsecond Ticker HAL API

> To enable tickless mode support in Mbed OS, you need to add the `MBED_TICKLESS` macro in the macros option of the target's section in the `targets.json` file.

Reviewer 2: Enable support is not the same as enable the mode. Which did we want?

> To enable tickless mode support in Mbed OS, you need to add the `MBED_TICKLESS` macro in the macros option of the target's section in the `targets.json` file.

Reviewer 2: ```suggestion
To enable tickless mode for your target, add the `MBED_TICKLESS` macro in target.json > your target's section > macro options.
```

> When the tickless mode is enabled, the device default SysTick mechanism overrides by the [OsTimer](../mbed-os-api-doxy/structos__timer__def.html) implementation based on either the [low power ticker](../mbed-os-api-doxy/group__hal__lp__ticker.html) or [microsecond ticker](../mbed-os-api-doxy/group__hal__us__ticker.html).

Reviewer 2: ```suggestion
When the tickless mode is enabled, the device's default SysTick mechanism overrides the [OsTimer](../mbed-os-api-doxy/structos__timer__def.html) implementation, based on either the [low power ticker](../mbed-os-api-doxy/group__hal__lp__ticker.html) or [microsecond ticker](../mbed-os-api-doxy/group__hal__us__ticker.html).
```

> When the tickless mode is enabled, the device suspends the kernel effectively disabling the systick and sets up either the [low power ticker](../mbed-os-api-doxy/group__hal__lp__ticker.html) or [microsecond ticker](../mbed-os-api-doxy/group__hal__us__ticker.html) through [mbed::internal::OsTimer](../mbed-os-api-doxy/structos__timer__def.html) based on target configuration.

Reviewer: Bit wary about mentioning `mbed::internal::` anything in documentation, but I guess it's a potentially useful detail for anyone debugging/maintaining the code.

> By default, tickless mode uses the low-power ticker if available. If a target's low-power ticker has an excessively long wake-up time or other performance issues, you must set `tickless-from-us-ticker` to `true` in the target's section in the `targets.json` file to make it use the microsecond ticker instead.

Reviewer 2: Low-power ticker, low power ticker or LowPowerTicker?

> By default, tickless mode uses the low-power ticker if available. If a target's low-power ticker has an excessively long wake-up time or other performance issues, you must set `tickless-from-us-ticker` to `true` in the target's section in the `targets.json` file to make it use the microsecond ticker instead.

Reviewer 2: ```suggestion
By default, tickless mode uses the low-power ticker (if available). If a target's low-power ticker has an excessively long wake-up time or other performance issues, make it use the microsecond ticker instead: set `tickless-from-us-ticker` to `true` in target.json > your target's section.
```

> If tickless mode uses the microsecond ticker, the device will only enter sleep rather than deep sleep, but will still avoid unnecessary tick calls.

Reviewer 2: ```suggestion
If tickless mode uses the microsecond ticker, the device will enter sleep rather than deep sleep, but will still avoid unnecessary tick calls.
```

> The expected scheduling for the same use-case as previously described should look like :

Reviewer 2: ```suggestion
The expected scheduling for the previous use case should look like:
```

> *Note that using the tickless mode prevents the user from using the associated timer.*

Reviewer: Oh, never spotted this sentence. Where did it come from?

It's not true in any meaningful sense for users. There is one HAL lpticker and one HAL usticker, yes, but they are always shared by the "ticker_api" framework. The OS timing uses that same sharing mechanism along with `Ticker`, `LowPowerTimeout` etc, and doesn't interfere with their use.

You couldn't use `LowPowerTicker` or `MBED_TICKLESS` at the same time as directly calling the HAL lpticker interrupt APIs yourself, but no-one would ever attempt that, I hope.

> ![](./resources/Tickless.png)

Reviewer 2: ```suggestion
![](resources/Tickless.png)
```

## Testing

> There are no dedicated tests validating tickless mode. Running all Mbed OS tests suites, with particular focus on HAL sleep and HAL low power ticker tests, provides sufficient coverage.

Reviewer 2: ```suggestion
There are no dedicated tests validating tickless mode. Running all Mbed OS test suites, with particular focus on HAL sleep and HAL low power ticker tests, provides sufficient coverage.
```

## References

> You can find more details on sleep modes in mbed in the section : [Mbed OS power management sleep](../../../apis/power-management-sleep.html)

Reviewer 2: ```suggestion
You can find more details on sleep modes in mbed in the section : [Mbed OS power management (sleep)](../apis/power-management-sleep.html)
```

# Tickless mode

Tickless mode is an optimization mechanism available in RTOS for suspending the SysTick. You can use it in situations when RTOS is idle for multiple ticks, so you can save power by entering uninterrupted sleep.

## Scheduling and sleep modes in Mbed OS

Mbed OS uses the SysTick timer in periods of 1ms to process threads' scheduling.

For instance, a system running two threads would see this timing:

![](resources/Normal_Tick.png)

Note that the device never enters deep sleep and wastes cycles in SysTick while all threads are asleep.

## Tickless mode

To support tickless mode in Mbed OS, your target needs to meet two requirements:

- Support for Sleep HAL API.
- Support for either Low Power or Microsecond Ticker HAL API.

To enable tickless mode for your target, add the `MBED_TICKLESS` macro in `target.json` (in your target's section):

```json
"macros_add": [
    "MBED_TICKLESS"
]
```

> When tickless mode is enabled, Mbed OS's default [OsTimer](../mbed-os-api-doxy/structos__timer__def.html) based on the [low power ticker](../mbed-os-api-doxy/group__hal__lp__ticker.html) replaces SysTick. If a target's low power ticker has an excessively long wake-up time or other performance issues, make it use the [microsecond ticker](../mbed-os-api-doxy/group__hal__us__ticker.html) instead by adding the below configuration in 'target.json' (in your target's section). And this configuration is not required for the target which does not have low power ticker as the default OsTimer based on the microsecond ticker.

Reviewer 2: ```suggestion
When tickless mode is enabled, Mbed OS's default [OsTimer](../mbed-os-api-doxy/structos__timer__def.html), based on the [low power ticker](../mbed-os-api-doxy/group__hal__lp__ticker.html), replaces SysTick. If a target's low power ticker has an excessively long wake-up time or other performance issues, make it use the [microsecond ticker](../mbed-os-api-doxy/group__hal__us__ticker.html) instead, by adding the `tickless-from-us-ticker` configuration in 'target.json' (in your target's section). This configuration is not required for a target that does not have low power ticker, because it will default to the microsecond ticker.
```

# Tickless mode

Tickless mode is an optimization mechanism available in RTOS for suspending the SysTick. You can use it in situations when RTOS is idle for multiple ticks, so you can save power by entering uninterrupted sleep. 

## Scheduling and sleep modes in Mbed OS

Mbed OS uses the SysTick timer in periods of 1ms to process threads' scheduling.

For instance, a system running two threads would see this timing:

![](resources/Normal_Tick.png)

Note that the device never enters deep sleep and wastes cycles in SysTick while all threads are asleep.

## Tickless mode

To support tickless mode in Mbed OS, your target needs to meet two requirements:

- Support for Sleep HAL API.
- Support for either Low Power or Microsecond Ticker HAL API.

To enable tickless mode for your target, add the `MBED_TICKLESS` macro in `target.json` (in your target's section):

```json
"macros_add": [
    "MBED_TICKLESS"
]
```

> When tickless mode is enabled, the device's default SysTick mechanism overrides by the [OsTimer](../mbed-os-api-doxy/structos__timer__def.html) implementation, based on either the [low power ticker](../mbed-os-api-doxy/group__hal__lp__ticker.html) or [microsecond ticker](../mbed-os-api-doxy/group__hal__us__ticker.html).

Reviewer 2: "overrides by the" is wrong - did you mean "overrides the OsTimer" or "overridden by the OsTimer"?

Reviewer: "is overridden by the OsTimer". But "replaced" would be better, I think.

And it's not the "device's default mechanism" - it's Mbed OS's default mechanism (if you don't define MBED_TICKLESS).

The device's default mechanism will be determined by whether or not its target.json defines MBED_TICKLESS. An application can override that in their own mbed_app.json (using `macros_add` or `macros_remove`).

Reviewer 2: ```suggestion
When tickless mode is enabled, Mbed OS's default [OsTimer](../mbed-os-api-doxy/structos__timer__def.html) replaces SysTick, based on either the [low power ticker](../mbed-os-api-doxy/group__hal__lp__ticker.html) or [microsecond ticker](../mbed-os-api-doxy/group__hal__us__ticker.html).
```

Reviewer 2: I think this needs to be broken. The bit about the lower power/microsecond ticker should be in the next paragraph, where we explain the default and how to override it.

Reviewer 2: As for the device default - should we clarify line 22?
